tests/sample/reflect_packet.c

```diff
@@ -100,6 +86,166 @@ reflect_packet(xdp_md_t* ctx)
+
+    // The inner IPv4 header will be after the old etherenet header.
+    next_header = (char*)(old_ethernet_header + 1);
+    if ((char*)next_header + sizeof(IPV4_HEADER) > (char*)ctx->data_end)
+        goto Done;
+    IPV4_HEADER* inner_ipv4_header = (IPV4_HEADER*)next_header;
+
+    // Copy over the old ethernet header to the new one.
+    memcpy(new_ethernet_header, old_ethernet_header, sizeof(ETHERNET_HEADER));
+    // Swap the MAC addresses.
+    swap_mac_addresses(new_ethernet_header);
+
+    // Copy over the inner IP header to the encap IP header.
+    memcpy(encap_ipv4_header, inner_ipv4_header, sizeof(IPV4_HEADER));
+    // Swap the IP addresses.
+    swap_ipv4_addresses(encap_ipv4_header);
+    // Adjust header fields.
+    encap_ipv4_header->Protocol = IPPROTO_IPV4;
+    encap_ipv4_header->HeaderLength = sizeof(IPV4_HEADER) / sizeof(uint32_t);
+    encap_ipv4_header->TotalLength = htons((ntohs(inner_ipv4_header->TotalLength) + sizeof(IPV4_HEADER)));
+    encap_ipv4_header->HeaderChecksum = 0;
+
+    rc = XDP_TX;
+
+Done:
+    return rc;
+}
+
+int
+encap_ipv6_reflect_packet(xdp_md_t* ctx)
+{
+    int rc = XDP_DROP;
+
+    // Adjust XDP context to allocate space for encap IPv6 header.
+    if (bpf_xdp_adjust_head(ctx, -sizeof(IPV6_HEADER)) < 0)
+        goto Done;
+
+    // The new ethernet header will be at the beginning of the expanded buffer.
+    char* next_header = (char*)ctx->data;
+    if ((char*)next_header + sizeof(ETHERNET_HEADER) > (char*)ctx->data_end)
+        goto Done;
+    ETHERNET_HEADER* new_ethernet_header = (ETHERNET_HEADER*)next_header;
+
+    // The old ethernet header is at an offset sizeof(IPV6_HEADER) from the start of the XDP buffer.
+    next_header = (char*)ctx->data;
+    if ((char*)next_header + sizeof(IPV6_HEADER) > (char*)ctx->data_end)
+        goto Done;
+    next_header = (char*)ctx->data + sizeof(IPV6_HEADER);
+    ETHERNET_HEADER* old_ethernet_header = (ETHERNET_HEADER*)next_header;
+
+    // The encap IPv4 header will be after the new etherenet header.
+    next_header = (char*)(new_ethernet_header + 1);
+    if ((char*)next_header + sizeof(IPV6_HEADER) > (char*)ctx->data_end)
+        goto Done;
+    IPV6_HEADER* encap_ipv6_header = (IPV6_HEADER*)next_header;
+
+    // The inner IPv6 header will be after the old etherenet header.
+    next_header = (char*)(old_ethernet_header + 1);
+    if ((char*)next_header + sizeof(IPV6_HEADER) > (char*)ctx->data_end)
+        goto Done;
+    IPV6_HEADER* inner_ipv6_header = (IPV6_HEADER*)next_header;
+
+    // Copy over the old ethernet header to the new one.
+    memcpy(new_ethernet_header, old_ethernet_header, sizeof(ETHERNET_HEADER));
+    // Swap the MAC addresses.
+    swap_mac_addresses(new_ethernet_header);
+
+    // Copy over the inner IP header to the encap IP header.
+    memcpy(encap_ipv6_header, inner_ipv6_header, sizeof(IPV6_HEADER));
+    // Swap the IP addresses.
+    swap_ipv4_addresses(encap_ipv6_header);
```

Comment:
```suggestion
    swap_ipv6_addresses(encap_ipv6_header);
```
BUG!

Reply:
wow - not sure how this compiled!! Will fix.